SSA/Core/HVector.lean

```diff
@@ -67,6 +67,9 @@ abbrev toSingle : HVector A [a₁] → A a₁ := toTuple
 abbrev toPair   : HVector A [a₁, a₂] → A a₁ × A a₂ := toTuple
 abbrev toTriple : HVector A [a₁, a₂, a₃] → A a₁ × A a₂ × A a₃ := toTuple
 
+abbrev fst : HVector A [a₁, a₂] → A a₁ := fun x => (toTuple x).fst
+abbrev snd : HVector A [a₁, a₂] → A a₂ := fun x => (toTuple x).snd
+
```

Comment:
If we have a `GetElem` instance then accessing individual elements becomes more consistent, regardless of the size of the vector (at the cost of maybe a little less type-safety, since we would no longer assert that the vector is exactly of size 2 when passing arguments to, e.g., `LLVM.and`, but I would say the consistency is worth it). Having `HVector.fst` hard-coded for a 2-element vector is a bit surprising.

Reply:
I can write this as:

```lean
@[simp]
def Op.denote (o : Op) (arg : HVector TyDenote.toType (OpSignature.sig o)) :
    (TyDenote.toType <| OpSignature.outTy o) :=
  match o with
  | Op.const _ val => const? val
  | Op.and _ => LLVM.and
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.or _ => LLVM.or
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.xor _ => LLVM.xor
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.shl _ => LLVM.shl
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.lshr _ => LLVM.lshr
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.ashr _ => LLVM.ashr
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.sub _ => LLVM.sub
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.add _ => LLVM.add
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.mul _ => LLVM.mul
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.sdiv _ => LLVM.sdiv
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.udiv _ => LLVM.udiv
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.urem _ => LLVM.urem
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.srem _ => LLVM.srem
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
  | Op.not _ => LLVM.not
                        (arg.get ⟨0, by simp [List.length]⟩)
  | Op.copy _ =>
                        (arg.get ⟨0, by simp [List.length]⟩)
  | Op.neg _ => LLVM.neg
                        (arg.get ⟨0, by simp [List.length]⟩)
  | Op.select _ =>
    select?
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
                        (arg.get ⟨2, by simp [List.length]⟩)
  | Op.icmp c _ => LLVM.icmp c
                        (arg.get ⟨0, by simp [List.length]⟩)
                        (arg.get ⟨1, by simp [List.length]⟩)
```

, but do not know how to shorten this.

I looked into `GetElem`, but https://lean-lang.org/functional_programming_in_lean/type-classes/indexing.html#overloading-indexing indicates the following:

> For the sake of flexiblity, GetElem has four parameters:
> 
> The type of the collection
> The type of the index
> The type of elements that are extracted from the collection
> A function that determines what counts as evidence that the index is in bounds

I am afraid I am not sure how to supply `The type of elements`

Reply:
Ah, sorry, I was too quick to suggest `GetElem`; that class does not really work for us because the return type of `get` depends on the particular index being accessed, while `GetElem` mandates a single `type of elements`, as you noted.

I guess I'm a bit concerned about special casing for pairs. The simplest generalization I see is to have `fst` and `snd` work so long as there are at least 1/2 elements, like:
```lean
abbrev fst : HVector A (a :: as) → A a                := head
abbrev snd : HVector A (a₁ :: a₂ :: as) → A a₂        := head ∘ tail
abbrev thd : HVector A (a₁ :: a₂ :: a₃ :: as) → A a₃  := head ∘ tail ∘ tail
```
This is still special casing in some sense, since we don't have a way to access element 4+, but we could in theory keep adding projections as needed (I imagine most operations wont have that many).

Alternatively, we can wrap `get` into our own method that indexes with a `Nat` and uses an auto-tactic for the proof obligations, as in
```lean
/-- `x.getN i` indexes into a `HVector` with a `Nat`,
using `get_elem_tactic` as an auto-tactic to prove that the index is within bounds -/
abbrev getN (x : HVector A as) (i : Nat) (hi : i < as.length := by get_elem_tactic) :
    A (as.get ⟨i, hi⟩) :=
  x.get ⟨i, hi⟩
```

This way, you can write, for example:
```lean
| Op.and _ => LLVM.and (op.getN 0) (op.getN 1)
```
The proof obligation is generally trivial, so `get_elem_tactic` should be able to work it out (and if not, we can extend it and/or use a different auto-tactic).

Reply:
Perfect. I introduced getN and cleaned up the code. Thank you for the suggestion. I did not know about auto-tactics.